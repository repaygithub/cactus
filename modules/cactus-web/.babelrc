{
  "presets": ["@repay/babel-preset"],
  "env": {
    "test": {
      "plugins": ["require-context-hook"]
    }
  }
}
