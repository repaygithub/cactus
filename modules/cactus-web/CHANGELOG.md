# v3.3.1 (Tue Nov 10 2020)

#### üêõ Bug Fix

- Fix polymophism behavior for `Button` (i.e. you can now use
`<Button as={Link} to="http://www.google.com">`) [#375](https://github.com/repaygithub/cactus/pull/375)
([@Dhalton](https://github.com/Dhalton))
- Fix issue where tooltip would cover Select dropdowns [#380](https://github.com/repaygithub/cactus/pull/380)
([@NicolasSimmonds](https://github.com/NicolasSimmonds))
- Fix disabled styles for the Toggle component [#384](https://github.com/repaygithub/cactus/pull/384)
([@NicolasSimmonds](https://github.com/NicolasSimmonds))

#### üè† Internal

- Refactor nav menu dropdowns [#383](https://github.com/repaygithub/cactus/pull/383)
([@wilysword](https://github.com/wilysword))

#### üî© Dependency Updates

- Upgrade `@repay/cactus-theme` to `v1.1.0`

#### Authors: 4

- Dhalton Huber ([@Dhalton](https://github.com/Dhalton))
- Glen Phelps ([@wilysword](https://github.com/wilysword))
- Nicolas Simmonds ([@NicolasSimmonds](https://github.com/NicolasSimmonds))
- Michael Soltow ([@mikesoltow](https://github.com/mikesoltow))
