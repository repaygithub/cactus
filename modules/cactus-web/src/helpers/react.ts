import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
