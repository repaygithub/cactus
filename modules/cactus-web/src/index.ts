export { AccessibleField, useAccessibleField } from './AccessibleField/AccessibleField'
export { Accordion } from './Accordion/Accordion'
export { default as ActionBar } from './ActionBar/ActionBar'
export { default as ActionProvider, useAction, useActionBarItems } from './ActionBar/ActionProvider'
export { Alert } from './Alert/Alert'
export { Avatar } from './Avatar/Avatar'
export { Breadcrumb } from './Breadcrumb/Breadcrumb'
export { Button } from './Button/Button'
export { Box } from './Box/Box'
export { Card } from './Card/Card'
export { CheckBox } from './CheckBox/CheckBox'
export { CheckBoxField } from './CheckBoxField/CheckBoxField'
export { default as CheckBoxGroup } from './CheckBoxGroup/CheckBoxGroup'
export { ColorPicker } from './ColorPicker/ColorPicker'
export { DataGrid } from './DataGrid/DataGrid'
export { DateInput } from './DateInput/DateInput'
export { DateInputField } from './DateInputField/DateInputField'
export { default as Dimmer } from './Dimmer/Dimmer'
export { FieldWrapper } from './FieldWrapper/FieldWrapper'
export { FileInput } from './FileInput/FileInput'
export { FileInputField } from './FileInputField/FileInputField'
export { Flex } from './Flex/Flex'
export { default as Grid } from './Grid/Grid'
export { default as usePopup } from './helpers/usePopup'
export { IconButton } from './IconButton/IconButton'
export { Label } from './Label/Label'
export { default as Layout } from './Layout/Layout'
export { Link } from './Link/Link'
export { default as List } from './List/List'
export { MenuBar } from './MenuBar/MenuBar'
export { MenuButton } from './MenuButton/MenuButton'
export { Modal } from './Modal/Modal'
export { Notification } from './Notification/Notification'
export { Pagination } from './Pagination/Pagination'
export { Preview } from './Preview/Preview'
export { PrevNext } from './PrevNext/PrevNext'
export { RadioButton } from './RadioButton/RadioButton'
export { RadioButtonField } from './RadioButtonField/RadioButtonField'
export { default as RadioGroup } from './RadioGroup/RadioGroup'
export {
  ScreenSizeContext,
  ScreenSizeProvider,
  SIZES,
  useScreenSize,
} from './ScreenSizeProvider/ScreenSizeProvider'
export { Select } from './Select/Select'
export { SelectField } from './SelectField/SelectField'
export { Spinner } from './Spinner/Spinner'
export { SplitButton } from './SplitButton/SplitButton'
export { default as StatusMessage } from './StatusMessage/StatusMessage'
export { StepAvatar } from './StepAvatar/StepAvatar'
export { StyleProvider } from './StyleProvider/StyleProvider'
export { Tab, TabController, TabList, TabPanel } from './Tabs/Tabs'
export { Table } from './Table/Table'
export { Text, Span } from './Text/Text'
export { TextArea } from './TextArea/TextArea'
export { TextAreaField } from './TextAreaField/TextAreaField'
export { TextButton } from './TextButton/TextButton'
export { TextInput } from './TextInput/TextInput'
export { TextInputField } from './TextInputField/TextInputField'
export { Toggle } from './Toggle/Toggle'
export { ToggleField } from './ToggleField/ToggleField'
export { Tooltip } from './Tooltip/Tooltip'
export { default as useId } from './helpers/useId'
export { ConfirmModal } from './ConfirmModal/ConfirmModal'
export { Tag } from './Tag/Tag'
export { Footer } from './Footer/Footer'
export { default as BrandBar } from './BrandBar/BrandBar'
export { Divider } from './Divider/Divider'
export { Header } from './Header/Header'
