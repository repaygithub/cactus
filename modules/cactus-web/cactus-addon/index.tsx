import cactusTheme, { ColorStyle, generateTheme } from '@repay/cactus-theme'
import addons, { makeDecorator } from '@storybook/addons'
import * as React from 'react'
import styled, { CSSObject } from 'styled-components'

import { StyleProvider } from '../src/StyleProvider/StyleProvider'
import {
  BACKGROUND_CHANGE,
  BORDER_BOX_CHANGE,
  DECORATOR_LISTENING,
  NAME,
  PROP_NAME,
  THEME_CHANGE,
} from './constants'

type AlignmentTypes = 'center' | 'left' | 'right' | 'bottom' | 'top'

interface StyledContainerBaseProps {
  inverse: boolean
  className?: string
  align?: AlignmentTypes
  overrides?: CSSObject
  borderBox?: boolean
}

const StyledContainerBase: React.FC<StyledContainerBaseProps> = ({
  className,
  children,
}): React.ReactElement => <div className={className}>{children}</div>

const alignmentMap: { [k in AlignmentTypes]: CSSObject } = {
  center: { justifyContent: 'center', alignItems: 'center' },
  left: { justifyContent: 'start', alignItems: 'center' },
  right: { justifyContent: 'end', alignItems: 'center' },
  bottom: { justifyContent: 'center', alignItems: 'end' },
  top: { justifyContent: 'center', alignItems: 'start' },
}

const StyledContainer = styled(StyledContainerBase)`
  position: relative;
  width: 100vw;
  height: 100vh;
  display: flex;
  overflow-y: auto;
  ${(p): ColorStyle => (p.inverse ? p.theme.colorStyles.base : p.theme.colorStyles.standard)};
  ${(p): CSSObject => alignmentMap[p.align]};
  ${(p): CSSObject => p.overrides}

  ${(p): string => p.borderBox && '* { box-sizing: border-box; }'}
`

StyledContainer.defaultProps = {
  align: 'center',
}

interface ProvideCactusThemeProps {
  channel: any
}

const ProvideCactusTheme: React.FC<ProvideCactusThemeProps> = ({
  channel,
  ...props
}): React.ReactElement => {
  const [theme, setTheme] = React.useState(cactusTheme)
  const [inverse, setInverse] = React.useState(false)
  const [borderBox, setBorderBox] = React.useState(false)

  React.useEffect((): (() => void) => {
    const updateTheme = (params: any): void => {
      setTheme(generateTheme(params))
    }

    const updateBackground = ({ inverse }: any): void => {
      setInverse(inverse)
    }

    const updateBorderBox = ({ borderBox }: any): void => {
      setBorderBox(borderBox)
    }

    channel.on(THEME_CHANGE, updateTheme)
    channel.on(BACKGROUND_CHANGE, updateBackground)
    channel.on(BORDER_BOX_CHANGE, updateBorderBox)

    channel.emit(DECORATOR_LISTENING)
    return (): void => {
      channel.removeListener(THEME_CHANGE, updateTheme)
      channel.removeListener(BACKGROUND_CHANGE, updateBackground)
      channel.removeListener(BORDER_BOX_CHANGE, updateBorderBox)
    }
  }, [channel])

  return (
    <StyleProvider theme={theme} global>
      <StyledContainer {...props} inverse={inverse} borderBox={borderBox} />
    </StyleProvider>
  )
}

export interface CactusAddonsOptions {
  align?: AlignmentTypes
  overrides?: CSSObject
}

export default makeDecorator({
  name: NAME,
  parameterName: PROP_NAME,
  wrapper: (
    getStory: any,
    context: any,
    { parameters, options }: { parameters?: CactusAddonsOptions; options?: CactusAddonsOptions }
  ): React.ReactElement => {
    const channel = addons.getChannel()
    parameters = { ...options, ...parameters }

    return (
      <ProvideCactusTheme {...parameters} channel={channel}>
        {getStory(context)}
      </ProvideCactusTheme>
    )
  },
})
