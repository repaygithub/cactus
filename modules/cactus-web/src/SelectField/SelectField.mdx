---
name: SelectField
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import SelectField from './SelectField'
import cactusTheme from '@repay/cactus-theme'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# SelectField

The `SelectField` is a pre-built Select field control that requires a label and handles accessibility internally.

### Try it out

export const code = `<form>
  <SelectField label="What's that in the sky?" name="ufo" options={['bird', 'plane', 'superman']} />
</form>`

<LiveProvider code={code} scope={{ SelectField }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

The `SelectField` should be used instead of the more rudimentary `Select` where possible to receive the accessibility benefits.

Shorter text is prefered for option labels to keep the output from assistive technologies concise.

## Basic usage

`options` can be provided as an array of strings or as an array of objects in the shape of `{ label: string, value: string }`.

```jsx
import React, { useState, useCallback } from 'react'

export default () => {
  const [values, setValues] = useState({ ufo: undefined })
  const handleChange = useCallback(
    ({ target }) => {
      setValues(state => setValues({ ...state, [target.name]: target.value }))
    },
    [setValues]
  )
  const onSubmit = useCallback(
    event => {
      event.preventDefault()
      const data = { ...values }
      // send data to api
    },
    [values]
  )

  return (
    <form onSubmit={onSubmit}>
      <SelectField
        label="What's that in the sky?"
        name="ufo"
        options={['bird', 'plane', 'superman']}
        onChange={handleChange}
        value={values.ufo}
      />
    </form>
  )
}
```

## Properties

<PropsTable of={SelectField} />
