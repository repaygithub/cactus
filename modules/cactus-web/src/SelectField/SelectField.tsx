import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { margin, MarginProps, width, WidthProps } from 'styled-system'

import AccessibleField, { FieldProps } from '../AccessibleField/AccessibleField'
import { omitMargins } from '../helpers/omit'
import Select, { OptionType, SelectProps, SelectValueType } from '../Select/Select'
import { FieldOnChangeHandler } from '../types'

interface SelectFieldProps
  extends MarginProps,
    WidthProps,
    FieldProps,
    Omit<SelectProps, 'id' | 'onChange' | keyof MarginProps | keyof WidthProps> {
  options: (OptionType | string | number)[]
  className?: string
  id?: string
  multiple?: boolean
  onChange?: FieldOnChangeHandler<SelectValueType>
}

const SelectFieldBase: React.FC<SelectFieldProps> = (props): React.ReactElement => {
  const {
    className,
    id,
    name,
    label,
    labelProps,
    tooltip,
    success,
    warning,
    error,
    width,
    disabled,
    autoTooltip,
    ...rest
  } = omitMargins(props) as Omit<SelectFieldProps, keyof MarginProps>
  const [isOpen, setIsOpen] = React.useState(false)
  return (
    <AccessibleField
      disabled={disabled}
      className={className}
      id={id}
      name={name}
      label={label}
      labelProps={labelProps}
      tooltip={tooltip}
      success={success}
      warning={warning}
      error={error}
      width={width}
      autoTooltip={autoTooltip}
      isOpen={isOpen}
    >
      {({ fieldId, labelId, name, ariaDescribedBy, status, disabled }): React.ReactElement => (
        <Select
          {...rest}
          disabled={disabled}
          status={status}
          name={name}
          id={fieldId}
          aria-labelledby={labelId}
          aria-describedby={ariaDescribedBy}
          getSelectOpen={(e) => setIsOpen(e)}
        />
      )}
    </AccessibleField>
  )
}

export const SelectField = styled(SelectFieldBase)`
  position: relative;
  ${margin}
  ${width}

  ${Select} {
    width: 100%;
  }
`

// @ts-ignore
SelectField.propTypes = {
  label: PropTypes.node.isRequired,
  labelProps: PropTypes.object,
  name: PropTypes.string.isRequired,
  // @ts-ignore
  options: Select.propTypes.options,
  className: PropTypes.string,
  id: PropTypes.string,
  success: PropTypes.string,
  warning: PropTypes.string,
  error: PropTypes.string,
  tooltip: PropTypes.string,
  onChange: PropTypes.func,
}

export default SelectField
