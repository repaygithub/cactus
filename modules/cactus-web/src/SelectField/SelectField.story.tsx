import { array, boolean, select, text } from '@storybook/addon-knobs'
import { Meta } from '@storybook/react/types-6-0'
import { Page } from 'puppeteer'
import React from 'react'

import { SelectValueType } from '../Select/Select'
import SelectField from './SelectField'

export default {
  title: 'SelectField',
  component: SelectField,
} as Meta

export const BasicUsage = (): React.ReactElement => (
  <SelectField
    label={text('label', `What's that in the sky?`)}
    name={text('name', 'ufo')}
    options={array('options', ['bird', 'plane', 'superman'])}
    disabled={boolean('disabled', false)}
    tooltip={text('tooltip', 'Select what you think you see in the sky.')}
    success={text('success', '')}
    warning={text('warning', '')}
    error={text('error', '')}
    autoTooltip={boolean('autoTooltip', true)}
    comboBox={boolean('comboBox', false)}
    canCreateOption={boolean('canCreateOption', true)}
    disableTooltip={select('disableTooltip', [false, true, undefined], false)}
  />
)

BasicUsage.parameters = {
  knobs: { escapeHTML: false },
  beforeScreenshot: async (page: Page) => {
    await page.click('button[name="ufo"]')
  },
}

export const CustomStyles = (): React.ReactElement => (
  <SelectField
    label="Who ya gonna call?"
    name="when_there_is_something_strange"
    options={['Ray Parker Jr.', 'Me maybe?', 'Ghostbusters']}
    width={text('width', '')}
    margin={text('margin', '3')}
  />
)
CustomStyles.parameters = { storyshots: false }

export const ControlledForm = (): React.ReactElement => {
  const [value, setValue] = React.useState<SelectValueType>(null)
  return (
    <SelectField
      label={text('label', `What's that in the sky?`)}
      name={text('name', 'ufo')}
      options={array('options', ['bird', 'plane', 'superman'])}
      disabled={boolean('disabled', false)}
      tooltip={text('tooltip', 'Select what you think you see in the sky.')}
      success={text('success', '')}
      warning={text('warning', '')}
      error={text('error', '')}
      onChange={(e) => setValue(e.target.value)}
      value={value}
    />
  )
}

ControlledForm.parameters = { knobs: { escapeHTML: false } }
ControlledForm.parameters = { storyshots: false }
