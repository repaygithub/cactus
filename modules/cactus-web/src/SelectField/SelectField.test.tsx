import * as React from 'react'
import { cleanup, fireEvent, render } from 'react-testing-library'
import { StyleProvider } from '@repay/cactus-web'
import SelectField from './SelectField'

afterEach(cleanup)

function closest(el: HTMLElement, matcher: (el: HTMLElement) => boolean) {
  while (el && el !== document.body) {
    if (el && matcher(el)) {
      return el
    }
    el = el.parentNode as HTMLElement
  }
  return null
}

describe('component: SelectField', () => {
  test('minimal snapshot', () => {
    const { container } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="prevent-random-id"
          options={['basic', 'options']}
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('render with complex options', () => {
    const { container } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="prevent-random-id"
          options={[{ label: 'Complex', value: 'complex' }, { label: 'Options', value: 'options' }]}
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render a disabled SelectField', () => {
    const { container } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="prevent-random-id"
          options={[{ label: 'Complex', value: 'complex' }, { label: 'Options', value: 'options' }]}
          disabled
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render an attached label', () => {
    const { getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="prevent-random-id"
          options={['basic', 'options']}
          tooltip="This is some information"
        />
      </StyleProvider>
    )

    expect(getByLabelText('Requires a label')).not.toBeNull()
  })

  test('should render an attached tooltip', () => {
    const { getByText, getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="prevent-random-id"
          options={['basic', 'options']}
          tooltip="This is some information"
        />
      </StyleProvider>
    )

    let tooltipTextEl = getByText('This is some information')
    let tooltipEl = closest(tooltipTextEl, el => el.getAttribute('role') === 'tooltip')
    expect(tooltipEl).not.toBeNull()

    let select = getByLabelText('Requires a label')
    expect(select.getAttribute('aria-describedby')).toContain((tooltipEl as HTMLElement).id)
  })

  test('should render a provided placeholder', () => {
    const { getByText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          options={['basic', 'options']}
          placeholder="This is that placeholder"
        />
      </StyleProvider>
    )

    expect(getByText('This is that placeholder')).not.toBeNull()
  })

  test('should render an accessible success message', () => {
    const { container, getByText, getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="for-the-snap"
          options={['basic', 'options']}
          success="You're GREAT!"
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
    let statusEl = getByText(`You're GREAT!`) as HTMLElement
    let select = getByLabelText('Requires a label')
    expect(select.getAttribute('aria-describedby')).toContain(statusEl.id)
  })

  test('should render an accessible warning message', () => {
    const { container, getByText, getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="for-the-snap"
          options={['basic', 'options']}
          warning="You got a ridiculous warning"
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
    let statusEl = getByText(`You got a ridiculous warning`) as HTMLElement
    let select = getByLabelText('Requires a label')
    expect(select.getAttribute('aria-describedby')).toContain(statusEl.id)
  })

  test('should render an accessible error message', () => {
    const { container, getByText, getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="for-the-snap"
          options={['basic', 'options']}
          error="You've done this incorrectly"
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
    let statusEl = getByText(`You've done this incorrectly`) as HTMLElement
    let select = getByLabelText('Requires a label')
    expect(select.getAttribute('aria-describedby')).toContain(statusEl.id)
  })

  test('supports margin props', () => {
    const { container, getByText, getByLabelText } = render(
      <StyleProvider>
        <SelectField
          label="Requires a label"
          name="the-test-select-field"
          id="for-the-snap"
          options={['basic', 'options']}
          margin="8px"
        />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  describe('should accept form event', () => {
    test('onChange', () => {
      const onChange = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <SelectField
            label="Requires a label"
            name="the-test-select-field"
            id="for-the-snap"
            options={['basic', 'options']}
            onChange={onChange}
          />
        </StyleProvider>
      )
      let select = getByLabelText('Requires a label') as HTMLElement
      fireEvent.click(select)
      fireEvent.click(getByText('basic'))
      expect(onChange).toHaveBeenCalledWith('the-test-select-field', 'basic')
    })

    test('onBlur', async () => {
      const onBlur = jest.fn()
      const { getByLabelText, rerender, getByText } = render(
        <StyleProvider>
          <SelectField
            label="Requires a label"
            name="the-test-select-field"
            id="for-the-snap"
            options={['basic', 'options']}
            onBlur={onBlur}
          />
        </StyleProvider>
      )
      let select = getByLabelText('Requires a label') as HTMLElement
      fireEvent.blur(select)
      expect(onBlur).toHaveBeenCalledWith('the-test-select-field', null)
    })

    test('onFocus', async () => {
      const onFocus = jest.fn()
      const { getByLabelText, getByText } = render(
        <StyleProvider>
          <SelectField
            label="Requires a label"
            name="the-test-select-field"
            id="for-the-snap"
            options={['basic', 'options']}
            onFocus={onFocus}
          />
        </StyleProvider>
      )
      let select = getByLabelText('Requires a label') as HTMLElement
      fireEvent.focus(select)
      expect(onFocus).toHaveBeenCalledWith('the-test-select-field', null)
    })
  })
})
