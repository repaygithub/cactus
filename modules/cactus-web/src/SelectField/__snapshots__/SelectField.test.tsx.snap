// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: SelectField minimal snapshot 1`] = `
.c0 + .c0 {
  margin-top: 16px;
}

.c4 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c11 {
  font-size: 8px;
  vertical-align: middle;
}

.c1 {
  position: relative;
}

.c1 .c3 {
  display: block;
  box-sizing: border-box;
  padding-left: 16px;
  padding-right: 28px;
}

.c1 .sc-fWWZrV {
  position: absolute;
  right: 8px;
  top: 2px;
  font-size: 16px;
}

.c1 .sc-cbeQSR {
  margin-top: 4px;
}

.c9 {
  font-style: italic;
  color: hsl(200,9%,35%);
  font-size: 18px;
}

.c7 {
  position: relative;
  box-sizing: border-box;
  min-width: 194px;
  width: 100%;
  height: 32px;
  padding: 0 24px 0 16px;
  background-color: transparent;
  border-radius: 8px;
  border-width: 1px;
  border-style: solid;
  border-color: hsl(200,9%,35%);
  text-align: left;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
}

.c7:disabled {
  border-color: hsl(0,0%,70%);
  color: hsl(0,0%,70%);
  cursor: not-allowed;
}

.c7:disabled .c8 {
  color: hsl(0,0%,70%);
}

.c7::-moz-focus-inner {
  border: 0;
}

.c7:focus,
.c7[aria-expanded] {
  border-color: hsl(200,96%,35%);
}

.c7:focus .c10,
.c7[aria-expanded] .c10 {
  color: hsl(200,96%,35%);
}

.c7[aria-expanded] .c10 {
  -webkit-transform: rotate3d(1,0,0,180deg);
  -ms-transform: rotate3d(1,0,0,180deg);
  transform: rotate3d(1,0,0,180deg);
}

.c7 .c10 {
  position: absolute;
  right: 14px;
  top: 10px;
}

.c6 {
  max-width: 100%;
}

.c2 {
  position: relative;
}

.c2 .c5 {
  width: 100%;
}

@media screen and (min-width:1024px) {
  .c4 {
    font-size: 18px;
    line-height: 1.5;
  }
}

@media screen and (min-width:1024px) {
  .c9 {
    font-size: 18px;
  }
}

@media screen and (min-width:1024px) {

}

<div>
  <div
    class="c0 c1 c2"
  >
    <label
      class="c3 c4"
      for="prevent-random-id"
      id="prevent-random-id-label"
    >
      Requires a label
    </label>
    <div
      class="c5 c6"
    >
      <div
        tabindex="-1"
      >
        <button
          aria-haspopup="listbox"
          aria-labelledby="prevent-random-id-label"
          aria-multiselectable="false"
          class="c7"
          id="prevent-random-id"
          name="the-test-select-field"
          type="button"
        >
          <span
            class="c8 c9"
          >
            Select an option
          </span>
          <svg
            class="c10 c11"
            fill="currentcolor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
              transform="rotate(-90 12 12)"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
